### PR TITLE
Issue #368 topic headings

### PR DESCRIPTION
**Pull request by bensinjin**

### CODE REVIEW

**Comment by rasmus-storjohann-PG** [comment]
@tomy-pg I'd like you to get going on reviewing other's code, can you please take a first pass at this PR and flag any problems (code style, unclear names, functions that do too much). I haven't looked at this PR yet, but I will after you're done to see if you missed anything ;-)

**Comment by tomy-pg** [comment]
Looks good @bensinjin . Definitely learned a lot from reviewing this.

**Comment by cherrielam** [comment]
@bensinjin Headings should only be on the home screen, not on related topics.

**Comment by bensinjin** [comment]
> @bensinjin Headings should only be on the home screen, not on related topics.

Yep, I made that change this morning

**Comment by cherrielam** [comment]
@bensinjin cool cool I forgot to comment on the PR as you asked.

**Comment by rasmus-storjohann-PG** [comment]
So the "Things to do right away" section needs to be at the top of the list.

File: src/components/tasks/build_topic_list_items_with_headings.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import * as R from 'ramda';
+import { TaskListItem } from '../../selectors/tasks/task_list_item';
+import { TopicListHeading } from './topic_list_heading_component';
+
+export type TopicListItemOrHeading = TaskListItem | TopicListHeading;
```
Comment (rasmus-storjohann-PG) [inline src/components/tasks/build_topic_list_items_with_headings.tsx]: I'm not a great fan of type names with "or" in them. Better to use to the type name to highlight what the various kinds of items have in common. So here, there is a list of items, each item is either a heading or a topic. So how about
```
export type ListItem = TopicListItem | HeadingListItem;
```

File: src/components/tasks/build_topic_list_items_with_headings.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+export const buildTopicsListItemsWithHeadings = (topics: ReadonlyArray<TaskListItem>): ReadonlyArray<TopicListItemOrHeading> => (
+    flattenGroupedTopics(groupTopicsByExploreSection(topics))
+);
+
+type GroupedTaskListItems  = { readonly [heading: string]: ReadonlyArray<TaskListItem> };
```
Comment (rasmus-storjohann-PG) [inline src/components/tasks/build_topic_list_items_with_headings.tsx]: I think this is much more readable:
```ts
type GroupedTaskListItems  = { 
    readonly [heading: string]: ReadonlyArray<TaskListItem> 
};
```

File: src/components/tasks/build_topic_list_items_with_headings.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+                id: heading,
+                heading: heading,
+                icon: groupedTopics[heading][0].exploreSection.icon,
+            },
+            ...groupedTopics[heading],
+        ],
+        [],
+        R.keys(groupedTopics),
+    )
+);
```
Comment (rasmus-storjohann-PG) [inline src/components/tasks/build_topic_list_items_with_headings.tsx]: SRP applies here. Long functions should be avoided, and particularly in functional code like this, "long" means 6-7lines or something. This function is doing a flattening with the reduce, as well as some other data manipulation that is not very clear. Breaking that out into a separate function, perhaps used with `R.map()`, will likely make this code easier to follow.

File: src/components/tasks/build_topic_list_items_with_headings.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+export const buildTopicsListItemsWithHeadings = (topics: ReadonlyArray<TaskListItem>): ReadonlyArray<TopicListItemOrHeading> => (
+    flattenGroupedTopics(groupTopicsByExploreSection(topics))
+);
+
+type GroupedTaskListItems  = { readonly [heading: string]: ReadonlyArray<TaskListItem> };
+
+const groupTopicsByExploreSection = (topics: ReadonlyArray<TaskListItem>): GroupedTaskListItems => (
+    R.groupBy((topic: TaskListItem) => topic.exploreSection.name, topics)
+);
```
Comment (rasmus-storjohann-PG) [inline src/components/tasks/build_topic_list_items_with_headings.tsx]: So there are two ways of pulling the grouping out into its own function, either what you've done here, or you can have something like 
```ts
const topicName = (topic: TaskListItem) => topic.exploreSection.name;
```
and then at the call site you would get
```ts
flattenGroupedTopics(R.groupBy(topicName, topics))
```
Not sure which you would prefer.

File: src/components/tasks/build_topic_list_items_with_headings.tsx (hunk 3 above)
Comment (rasmus-storjohann-PG) [reply src/components/tasks/build_topic_list_items_with_headings.tsx]: Actually, now I look at it closer, it's actually very nice. But the way the code is broken up, the big picture of what you're doing is getting a bit lost. So your new accumulator is the old accumulator, plus a heading item, plus the group of topics belonging under that heading. You can take advantage of the fact that the first argument to `R.reduce()` is a function, so you can have local variables in there that you can give names that makes the flow much easier to follow. And you can pull the low level details of how a heading item is actually constructed into a helper function, making this function consistently high level and clearer. @tomy-pg and @DarwishMenna for visibility.

Comment (rasmus-storjohann-PG) [reply src/components/tasks/build_topic_list_items_with_headings.tsx]: Do you need the separate `id` and `heading` properties in the heading item?

File: src/components/tasks/topic_list_heading_component.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+export interface TopicListHeading {
+    readonly id: string;
+    readonly heading: string;
+    readonly icon: string;
+}
+
+export interface TopicListHeadingProps {
+    readonly heading: string;
+    readonly icon: string;
+}
```
Comment (rasmus-storjohann-PG) [inline src/components/tasks/topic_list_heading_component.tsx]: Do you need both of these types?

File: src/selectors/tasks/select_saved_tasks.ts [hunk 6]
```diff
@@ -1,12 +1,9 @@
...
 import { pickSavedTaskIds } from './pick_saved_task_ids';
 
-export const selectSavedTasks = (appStore: Store): ReadonlyArray<TaskListItem> => {
-    const savedTasksList = pickSavedTaskIds(appStore);
-    return savedTasksList.map((taskId: store.Id) => {
-        return selectTaskAsListItem(appStore, taskId);
-    });
-};
+export const selectSavedTasks = (appStore: Store): ReadonlyArray<TaskListItem> => (
+    R.map(buildSelectorTaskListItem(appStore), pickSavedTaskIds(appStore))
```
Comment (rasmus-storjohann-PG) [inline src/selectors/tasks/select_saved_tasks.ts]: Nice improvement

**Comment by bensinjin** [comment]
@bensinjin TODO: Recreate this PR against develop

**Comment by rasmus-storjohann-PG** [comment]
Closing PR against master, will open a new PR against develop

**Pull request closed without merging**